### PR TITLE
fix(47858): 'Any' on anonymous function argument response in generic

### LINKED ISSUE #47858: 'Any' on anonymous function argument response in generic
# Bug Report

### 🔎 Search Terms

* Anonymous function response

### 🕗 Version & Regression Information

4.5 and below

### ⏯ Playground Link

Simplified:

https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABFATgQzAZ2HFBbAHgHlEBTADylLABNNEBvRAbQGtSBPALkU1RjABzALo8AFADopaFIMw8MHZsICUiALwA+RGJmCeYEHgBGpFGq2IAbnBg1EAX01iBABxBQeRNQwBQvh39UDGxcPDE-RCjEYDAARnE3Dx4+FAFBC20xGjQoNEzGf2jHABpfYuAAJkSwd09efiECyOLolFIoEBQkbNz8jW0-ctaowOKHMoqwAGYXWuSGtKbCkaj2zu6dHLzmotWx6ImAlV8gA

Actual:

https://www.typescriptlang.org/play?useUnknownInCatchVariables=true&ts=4.5.4#code/C4TwDgpgBAsghiARhACgJwPYFsCWBnCAHgBUA+KAXimKgB8p1t8iyAoUSKAZWDmAgBmAVwA2AJQh5CXclXhJUmXAUIAKOGgDmALm4BKSuS50oANww4AJqXbhoPPoNEBBAMbAcGAHYBhNBD4MNGlZKFUAOkiNTTxdOC8QAG0AXQMKI15+YXFJENZWHC9+NAE4V2g3D28uCGAQqABvVigWqESAawgQXTxgNELNZN0HLJd3T19-QOCZVgBffMLi0vKoSomACXjLEQhg5ygIAA9+L0s8KHik5PIm1rDI8OjYtb1dcyt5xaK9lYq8EBeVzrbxbM67faHE4QM4XK4pW7NVoRKJaF7ON4MJTMQgfaxfViuERwPAXFAYSzSAA0a0R9zA-VMjighRwHjgIhGEGGSJaDJwTP4l3G1Vqumc+Xurm8vTQQncQVUrPZnMy3O4NLKVS8NWAlEac0uFwxjV59ygwAAFvhwsqcByufq7Q61Wb7labVqJrr9V7RcAzQszd5CAB5KGnc5rEU62ohUiqDCIABWulDmPJlK4NIOADJGlA3a1EgBpFleKCdEAYARQUNDOul5IRmFRlFPNG6QoCPYMNLkbu9sRF81QAD8UDELdhWKYKiuNlHS9aE+cAKBIK8YJ2e0IKEXy6X4pj24he4Ph6guiEXnaXgwAHcvIG6Uv-MAhGgK14IA+GBTVA9PBbS8Nl7VVRw9EDAlpS8XooDQfUfz-TNVAaKBpRvYBdAARgABigOYoPucJvDQkcjUBVwMMtCBXHaVQqx6PoBgMO5LygAB6TiMJlPUKyoOAHzgNlKKBHwORERiumIjj30-CtVDwftTQ481uKgYDMKKKAAGoqGfNSFmXOYqQov0vF0VRLKgLwhCwZA0BU9jL3kr8wmUwxVLU1oNK0jAsL0gyKPuYyl1M8zLEsKybLshy9hUpTnJC1p-MC-TbIosKWiI1ggA

### 💻 Code

```ts
function transform<O extends { [key: string]: (...args: any[]) => (arg: number) => void }>(input: O) {

}

transform({
    fn1: (input: string) => (data) => {

    },
    f2: (input: string) => {
        return (data) => { // coming back as any instead of number

        }
    },
    fn3(input: string) {
        return (data) => {

        }
    },
})

```

### 🙁 Actual behavior

The generic argument in the return function is coming back as `any` instead of `number` when it's returned as an anonymous inline function.

### 🙂 Expected behavior

Expecting the inline anonymous function response to be of type number.

### PR DESCRIPTION
**Pull request by chbdetta**
Fixes #47858

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -17041,8 +17041,13 @@ namespace ts {
         }
 
         function hasContextSensitiveReturnExpression(node: FunctionLikeDeclaration) {
-            // TODO(anhans): A block should be context-sensitive if it has a context-sensitive return value.
-            return !node.typeParameters && !getEffectiveReturnTypeNode(node) && !!node.body && node.body.kind !== SyntaxKind.Block && isContextSensitive(node.body);
+            // If one of the returned expressions is context sensitive, the body block is context sensitive
+            if (!node.typeParameters && !getEffectiveReturnTypeNode(node) && !!node.body) {
+                return node.body.kind !== SyntaxKind.Block ? isContextSensitive(node.body) :
```
Comment (RyanCavanaugh) [inline src/compiler/checker.ts:17046]: Nit: Instead of `x !== y ? a : b`, use `x === y ? b : a`

**Review by weswigham — CHANGES_REQUESTED** [review]
IIRC, one of the reasons we haven't actually done this on-the-surface-very-simple-TODO is because the `this` within the functions that becomes contextually typed after this has consistency/calculation issues (since the `this` is dependent on the containing object this is dependent on the function type which is dependent on the `this`...). Given that, some more test cases that use `this` in the function body (and the return expression itself) seem warranted.

Also cc @rbuckton who reviewed the original TODO comment PR and might remember more.

**Comment by sandersn** [comment]
Do you want to keep working on this? Otherwise I'd like to close it to reduce the number of open PRs.

**Comment by chbdetta** [comment]
@sandersn feel free to close this since I am not sure what action should be taken here

**Comment by weswigham** [comment]
> Given that, some more test cases that use this in the function body (and the return expression itself) seem warranted.

^^^

**Pull request closed without merging**